pkg/e2e/infra/infra.go

```diff
@@ -123,40 +122,90 @@ func invalidMachinesetWithEmptyProviderConfig() *mapiv1beta1.MachineSet {
+	o.Expect(err).ToNot(o.HaveOccurred())
+
+	providerSpec := workers[0].Spec.Template.Spec.ProviderSpec.DeepCopy()
+	clusterName := workers[0].Spec.Template.Labels[e2e.ClusterKey]
+
+	clusterInfra, err := e2e.GetInfrastructure(client)
+	o.Expect(err).NotTo(o.HaveOccurred())
+	o.Expect(clusterInfra.Status.InfrastructureName).ShouldNot(o.BeEmpty())
+
+	msName := e2e.RandomString(clusterInfra.Status.InfrastructureName)
+
+	return e2e.MachineSetParams{
+		Name:         msName,
+		Replicas:     int32(replicas),
+		ProviderSpec: providerSpec,
+		Labels: map[string]string{
+			"mhc.e2e.openshift.io": msName,
+			e2e.ClusterKey:         clusterName,
+		},
+	}
+}
+
+func deleteObject(client runtimeclient.Client, obj runtime.Object) error {
+	cascadeDelete := metav1.DeletePropagationForeground
+	return client.Delete(context.TODO(), obj, &runtimeclient.DeleteOptions{
+		PropagationPolicy: &cascadeDelete,
+	})
+}
+
+func deleteObjects(client runtimeclient.Client, delObjects map[string]runtime.Object) error {
+	// Remove resources
+	for _, obj := range delObjects {
+		if err := deleteObject(client, obj); err != nil {
+			glog.Errorf("[cleanup] error deleting object: %v", err)
+			return err
+		}
+	}
+	return nil
+}
+
 var _ = g.Describe("[Feature:Machines] Managed cluster should", func() {
 	defer g.GinkgoRecover()
 
-	deleteObject := func(client runtimeclient.Client, obj runtime.Object) error {
-		cascadeDelete := metav1.DeletePropagationForeground
-		return client.Delete(context.TODO(), obj, &runtimeclient.DeleteOptions{
-			PropagationPolicy: &cascadeDelete,
-		})
-	}
+	var client runtimeclient.Client
+	var machineSet *mapiv1beta1.MachineSet
+	var machineSetParams e2e.MachineSetParams
 
-	g.It("have machines linked with nodes", func() {
+	g.BeforeEach(func() {
 		var err error
-		client, err := e2e.LoadClient()
-		o.Expect(err).NotTo(o.HaveOccurred())
 
-		o.Expect(isOneMachinePerNode(client)).To(o.BeTrue())
-		o.Expect(err).NotTo(o.HaveOccurred())
+		client, err = e2e.LoadClient()
+		o.Expect(err).ToNot(o.HaveOccurred())
+
+		machineSetParams = buildMachineSetParams(client, 3)
+
+		g.By("Creating a new MachineSet")
+		machineSet, err = e2e.CreateMachineSet(client, machineSetParams)
+		o.Expect(err).ToNot(o.HaveOccurred())
+
+		e2e.WaitForMachineSet(client, machineSet.GetName())
+	})
+
+	g.AfterEach(func() {
+		g.By("Deleting the new MachineSet")
+		err := client.Delete(context.Background(), machineSet)
+		o.Expect(err).ToNot(o.HaveOccurred())
+
+		e2e.WaitForMachineSetDelete(client, machineSet)
 	})
 
 	g.It("have ability to additively reconcile taints from machine to nodes", func() {
-		var err error
-		client, err := e2e.LoadClient()
-		o.Expect(err).NotTo(o.HaveOccurred())
+		selector := machineSet.Spec.Selector
+		machines, err := e2e.GetMachines(client, &selector)
+		o.Expect(err).ToNot(o.HaveOccurred())
 
-		machines, err := e2e.GetMachines(client)
-		o.Expect(err).NotTo(o.HaveOccurred())
-		o.Expect(len(machines)).To(o.BeNumerically(">", 0))
```

Comment:
Might this check be worth keeping before accessing a potentially empty list on the next line? Alternatively `o.Expect(machines).ToNot(o.BeEmpty())` is probably a nicer way to do it

Reply:
:+1: